packages/dht/src/dht/PeerManager.ts

```diff
@@ -0,0 +1,320 @@
+                    contact.getPeerDescriptor(),
+                    this.neighborList!.getClosestContacts(this.config.getClosestContactsLimit).map((peer) => peer.getPeerDescriptor())
+                )
+            } else {    // open connection by pinging
+               
+                contact.ping().then((result) => {
+                    if (result) {
+                        logger.trace(`Added new contact ${contact.getPeerId().value.toString()}`)
+                        this.emit(
+                            'newKbucketContact',
+                            contact.getPeerDescriptor(),
+                            this.neighborList!.getClosestContacts(this.config.getClosestContactsLimit).map((peer) => peer.getPeerDescriptor())
+                        )
+                    } else {
+                        this.config.connectionManager?.weakUnlockConnection(contact.getPeerDescriptor())
+                        this.removeContact(contact.getPeerDescriptor())
+                        this.addClosestContactToBucket()
+                    }
+                    return
+                }).catch((_e) => {
+                    this.config.connectionManager?.weakUnlockConnection(contact.getPeerDescriptor())
+                    this.removeContact(contact.getPeerDescriptor())
+                    this.addClosestContactToBucket()
+                })
+            }
+        }
+    }
+
+    private addClosestContactToBucket(): void {
+        if (this.stopped) {
+            return
+        }
+        const closest = this.getClosestActiveContactNotInBucket()
+        if (closest) {
+            this.handleNewPeers([ closest.getPeerDescriptor() ])
+        }
+    }
+
+    private getClosestActiveContactNotInBucket(): RemoteDhtNode | undefined {
+        for (const contactId of this.neighborList!.getContactIds()) {
+            if (!this.bucket!.get(contactId.value) && this.neighborList!.isActive(contactId)) {
+                return this.neighborList!.getContact(contactId).contact
+            }
+        }
+        return undefined
+    }
+
+    public handleConnected(peerDescriptor: PeerDescriptor): void {
+        if (this.config.ownPeerId!.equals(PeerID.fromValue(peerDescriptor.kademliaId))) {
+            logger.error('handleConnected() to self')
+            return
+        }
+
+        const RemoteDhtNode = this.config.createRemoteDhtNode(peerDescriptor)
+        if (!this.connections.has(PeerID.fromValue(RemoteDhtNode.id).toKey())) {
+            this.connections.set(PeerID.fromValue(RemoteDhtNode.id).toKey(), RemoteDhtNode)
+        } else {
+            logger.trace('new connection not set to connections, there is already a connection with the peer ID')
+        }
+    }
+
+    public handleDisconnected(peerDescriptor: PeerDescriptor, disconnectionType: DisconnectionType): void {
+        this.connections.delete(keyFromPeerDescriptor(peerDescriptor))
+        
+        // only remove from bucket if we are on layer 0
```

Comment:
Maybe a comment _why_ it is removed only if we are on layer 0? Also the layer seems to affect also whether we call `this.removeContact(peerDescriptor, true)`